### ISSUE TITLE
Neovim crashes on opening lisp-file. Could not reference parinfer library.

### ISSUE THREAD

**Issue by duikboot** [eraserhd/parinfer-rust#6]
Thanks for your plugin, it's very useful for me.

I updated parinfer-rust, but now it crashes on opening a lisp file.

```

1186Cthread '<unnamed>' panicked at 'Could not reference parinfer library.', src/lib.rs:386:13                                                                                                                         note: Run with `RUST_BACKTRACE=1` for a backtrace.
fatal runtime error: failed to initiate panic, error 5
[1]    21581 abort (core dumped)  nvim cl-duikboot.asd
```
```
$ nvim --version
NVIM v0.2.3-874-g739fb93a9
Build type: RelWithDebInfo
LuaJIT 2.0.5
Compilation: /usr/bin/cc -Wconversion -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -O2 -g -DMIN_LOG_LEVEL=3 -Og -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fdiagnostics-color=auto -Wno-array-bounds -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -I/home/arjen/Programms/neovim/build/config -I/home/arjen/Programms/neovim/src -I/home/arjen/Programms/neovim/.deps/usr/include -I/usr/include -I/home/arjen/Programms/neovim/build/src/nvim/auto -I/home/arjen/Programms/neovim/build/include
Compiled by arjen@localhost.localdomain

```

**Comment by eraserhd** [comment]
That is... weird, indeed.  I thought parinfer and cparinfer were resolved at link time.

1. Try `cargo clean` and `cargo build --release` in the cparinfer subdirectory and see if it works.
2. If not, try `cargo build --release` in the parinfer directory, and then in the cparinfer directory. 
3. If on Mac OS, try the fix I'm about to merge.

Otherwise, what platform are you on?

Let me know what happens.
Thanks!

**Comment by duikboot** [comment]
Thanks for your fast reply.

1. It cleans and build and I get the crash mentioned above
2. error: could not find `Cargo.toml` in `/home/arjen/config/dotfiles/_neovim/bundle/parinfer-rust` or any parent directory
3. I'm on linux (Fedora 27)

If I run `git checkout 582e2fd` and do a clean and build, it works. I tried a lot commits (not all), but this one works for me.

Thanks

**Comment by eraserhd** [comment]
Re: 2, you should be in `parinfer-rust/parinfer` or `parinfer-rust/cparinfer` for `cargo clean` or `cargo build --release` to work...  But you are saying `cargo build --release` works on `582e2fd`?

**Comment by duikboot** [comment]
`cargo build --release` always worked, but even then, neovim crashes on opening a lisp file. 
I searched for the last version it worked. 
It is: `a9a6287`

**Comment by eraserhd** [comment]
Ahhh!  I see the problem, I think.

**Comment by eraserhd** [comment]
OK, I've added more info to the panic message: the file it is trying to reference and the dlopen() error info.  Can you try this and tell me what the panic message says now?

Thanks!

**Comment by duikboot** [comment]
Sure:
```
"src/cl-duikboot.lisp" 39L, 747Cthread '<unnamed>' panicked at 'Could not reference cparinfer library "/home/arjen/config/dotfiles/_neovim/bundle/parinfer-rust/cparinfer/target/release/libcparinfer.so": "/home/arjen/config/dotfiles/_neovim/bundle/parinfer-rust/cparinfer/target/release/libcparinfer.so: invalid mode for dlopen(): Invalid argument".', src/lib.rs:387:13                                                                                                            
note: Run with `RUST_BACKTRACE=1` for a backtrace.
                                                                                                                                                                                    fatal runtime error: failed to initiate panic, error 5
                                                                                                                                                                                                                                          [1]    15048 abort (core dumped)  vim src/cl-duikboot.lisp

```

**Comment by eraserhd** [comment]
Interesting.  Does your manpage for `dlopen()` have `RTLD_NOLOAD`, `RTLD_GLOBAL`, and `RTLD_NODELETE`?

**Comment by duikboot** [comment]
Yes it does, but it says the following in the manpage:

```
BUGS
       As  at glibc 2.24, specifying the RTLD_GLOBAL flag when calling dlmopen() generates an error.
Furthermore, specifying RTLD_GLOBAL when calling dlopen() results in a program crash (SIGSEGV) if the call is made from any object loaded in a namespace other than the initial namespace.
```

**Comment by eraserhd** [comment]
Haha.  OK, so the problem is that, with glibc, we are *required* to pass either RTLD_LAZY or RTLD_NOW.  This should be fixed in b3f9aea, which I just pushed.  Please try!

**Comment by duikboot** [comment]
YES! It is fixed. Thank you
